File: go/ast_shape_test.go
// Copyright (c) 2026 Richard Rodger and other contributors, MIT License

package tabnasexpr

import (
	"encoding/json"
	"testing"
)

// TestASTShapeDivergesFromTypeScript pins a DIVERGENCE, not a contract.
// See DIVERGENCE.md: this port's parsed AST serialises to a different
// shape from the TypeScript port's, and closing that is a breaking change
// to this port's public Go types.
//
// It is pinned on BOTH sides — the TypeScript twin is
// ts/test/ast-shape.test.ts — so the record cannot outlive the
// divergence: repairing either port turns that port's test red and forces
// the pair and the DIVERGENCE.md entry to be deleted together (ADR-14).
//
// Measured on `{a:1+2}`, the simplest expression that shows it:
//
//	TypeScript  a -> [ {src:"+", left:…, name:"addition-infix", …} ]
//	Go          a -> { Val: [ {Name:"addition-infix", Src:"+", …} ], … }
func TestASTShapeDivergesFromTypeScript(t *testing.T) {
	v, err := Parse("{a:1+2}")
	if err != nil {
		t.Fatalf("parse failed: %v", err)
	}
	raw, err := json.Marshal(v)
	if err != nil {
		t.Fatalf("marshal failed: %v", err)
	}

	var doc map[string]any
	if err := json.Unmarshal(raw, &doc); err != nil {
		t.Fatalf("unmarshal failed: %v", err)
	}
	a, ok := doc["a"]
	if !ok {
		t.Fatal("no `a` key, so this test proves nothing")
	}

	// 1. The wrapper. TypeScript yields the term list directly.
	obj, isObj := a.(map[string]any)
	if !isObj {
		t.Fatalf("`a` is %T, not an object — if the wrapper has been "+
			"removed to match TypeScript, delete this test, its twin in "+
			"ts/test/ast-shape.test.ts, and the DIVERGENCE.md entry", a)
	}
	terms, hasVal := obj["Val"]
	if !hasVal {
		t.Error("`a` no longer carries Val — see the note above")
	}

	// 2. The naming. No json tags, so encoding/json emits the exported
	// Go names where TypeScript emits lower-case ones.
	list, isList := terms.([]any)
	if !isList || 0 == len(list) {
		t.Fatalf("Val is %T with no terms, so the naming check below "+
			"would pass vacuously", terms)
	}
	op, isOp := list[0].(map[string]any)
	if !isOp {
		t.Fatalf("first term is %T, not an object", list[0])
	}
	if _, pascal := op["Name"]; !pascal {
		t.Error("the first term has no `Name` key — if json tags have " +
			"been added to match TypeScript's `name`, delete this test, " +
			"its twin, and the DIVERGENCE.md entry together")
	}
	if _, lower := op["name"]; lower {
		t.Error("the first term has a lower-case `name` key, so this port " +
			"now agrees with TypeScript — see the note above")
	}
}

File: ts/test/ast-shape.test.ts
/* Copyright (c) 2026 Richard Rodger and other contributors, MIT License */

import { describe, test } from 'node:test'
import assert from 'node:assert'

import { Tabnas } from '@tabnas/parser'
import { jsonic } from '@tabnas/jsonic'

import { Expr } from '..'

// The TypeScript twin of go/ast_shape_test.go. Together they pin a
// DIVERGENCE, not a contract — see DIVERGENCE.md: the two ports serialise
// the parsed AST to different shapes, and closing that is a breaking
// change to the Go port's public types.
//
// Pinned on BOTH sides so the record cannot outlive what it records:
// repairing either port turns that port's test red and forces the pair
// and the DIVERGENCE.md entry to be deleted together (ADR-14).
//
// Measured on `{a:1+2}`, the simplest expression that shows it:
//
//   TypeScript  a -> [ {src:"+", left:…, name:"addition-infix", …} ]
//   Go          a -> { Val: [ {Name:"addition-infix", Src:"+", …} ], … }
describe('AST shape', () => {
  test('this port yields the term list directly, where Go wraps it', () => {
    const tn = new Tabnas().use(jsonic).use(Expr)
    const out: any = tn.parse('{a:1+2}')

    const a = out?.a
    assert.notEqual(a, undefined, 'no `a` key, so this test proves nothing')

    // 1. No wrapper. Go yields { Val: [...], Child, Implicit, Meta }.
    assert.ok(Array.isArray(a),
      '`a` is not an array — if this port has grown Go\'s wrapper, delete ' +
      'this test, its twin in go/ast_shape_test.go, and the ' +
      'DIVERGENCE.md entry together')
    assert.notEqual(a.length, 0,
      'no terms, so the naming check below would pass vacuously')

    // 2. Lower-case naming. Go's structs carry no json tags, so it emits
    // the exported Go names.
    const op = a[0]
    assert.ok('name' in op,
      'the first term has no `name` key — see the note above')
    assert.equal('Name' in op, false,
      'the first term has a PascalCase `Name` key, so this port now ' +
      'matches Go — see the note above')
  })
})
